src/realm/array_flex.cpp

```diff
@@ -175,42 +175,73 @@ bool ArrayFlex::find_all(const Array& arr, int64_t value, size_t start, size_t e
-            if constexpr (v == false)
-                return find_all_fields_unsigned_GE(MSBs, a, b);
-            REALM_UNREACHABLE();
-        }
+    if constexpr (std::is_same_v<Cond, Equal>)
+        return find_all_fields_EQ(MSBs, a, b);
+    else if constexpr (std::is_same_v<Cond, NotEqual>)
+        return find_all_fields_NE(MSBs, a, b);
+    else if constexpr (std::is_same_v<Cond, GreaterEqual>) {
+        if constexpr (std::is_same_v<Type, ArrayFlex::WordTypeValue>)
+            return find_all_fields_signed_GE(MSBs, a, b);
+        if constexpr (std::is_same_v<Type, ArrayFlex::WordTypeIndex>)
+            return find_all_fields_unsigned_GE(MSBs, a, b);
+        REALM_UNREACHABLE();
+    }
+
+    else if constexpr (std::is_same_v<Cond, Greater>)
+        return find_all_fields_signed_GT(MSBs, a, b);
+    else if constexpr (std::is_same_v<Cond, Less>)
+        return find_all_fields_unsigned_LT(MSBs, a, b);
+}
+
+template <typename Type>
+inline uint64_t get_MSBs(const Array& arr)
+{
+    if constexpr (std::is_same_v<Type, ArrayFlex::WordTypeIndex>)
+        return arr.get_encoder().ndx_msb();
+    return arr.get_encoder().msb();
+}
+
+template <typename Type>
+inline size_t get_field_count(const Array& arr)
+{
+    if constexpr (std::is_same_v<Type, ArrayFlex::WordTypeIndex>)
+        return arr.get_encoder().ndx_field_count();
+    return arr.get_encoder().field_count();
+}
+
+template <typename Type>
+inline size_t get_bit_count_per_iteration(const Array& arr)
+{
+    if constexpr (std::is_same_v<Type, ArrayFlex::WordTypeIndex>)
+        return arr.get_encoder().ndx_bit_count_per_iteration();
+    return arr.get_encoder().bit_count_per_iteration();
+}
 
-        else if constexpr (std::is_same_v<Cond, Greater>)
-            return find_all_fields_signed_GT(MSBs, a, b);
-        else if constexpr (std::is_same_v<Cond, Less>)
-            return find_all_fields_unsigned_LT(MSBs, a, b);
-    };
+inline uint64_t last_word_mask(size_t total_bit_count_left)
+{
+    // generates a mask of 1s shifting by 64-<bit count left> 0xFFFFULL.
+    // Useful for extracting the last number of bits from a vector of matching positions.
+    return 0xFFFFFFFFFFFFFFFFULL >> (64 - total_bit_count_left);
+}
 
+template <typename Cond, typename Type>
+inline size_t ArrayFlex::parallel_subword_find(const Array& arr, size_t offset, size_t width, size_t start,
+                                               size_t end, uint64_t search_vector, int64_t total_bit_count_left) const
+{
+    const auto MSBs = get_MSBs<Type>(arr);
+    const auto field_count = get_field_count<Type>(arr);
+    const auto bit_count_pr_iteration = get_bit_count_per_iteration<Type>(arr);
+    REALM_ASSERT_DEBUG(total_bit_count_left >= 0);
     unaligned_word_iter it((uint64_t*)(arr.m_data), offset + start * width);
     uint64_t vector = 0;
-    while (total_bit_count_left >= bit_count_pr_iteration) {
+    while (total_bit_count_left >= (int64_t)bit_count_pr_iteration) {
```

Comment:
forcing 64 bit representation here is not needed and just makes it slower on 32 bit machines.